File: examples/MovingAverage/client/rb/AdvancedStatic/MovingAverageClient.rb
#!/usr/bin/env ruby

$LOAD_PATH.push('../gen-rb')

require 'thrift'
require 'moving_average_service'

def check(data_in, data_out, size)
  status = 0
  (1..(size - 2)).each do |i|
    next if (((data_in[i - 1] + data_in[i] + data_in[i + 1]) / 3) -
            data_out[i]).abs < 0.0001
    puts "Output data @ #{i} =  #{data_out[i]} "\
         "(expected #{(data_in[i - 1] + data_in[i] + data_in[i + 1]) / 3})"
    status += 1
  end

  status
end

begin

  include Com::Maxeler::MovingAverage

  start_time = Time.now
  start_dfe_time = start_time

  port = 9090

  # Make socket
  socket = Thrift::Socket.new('localhost', port)

  # Buffering is critical. Raw sockets are very slow
  transport = Thrift::BufferedTransport.new(socket)

  # Wrap in a protocol
  protocol = Thrift::BinaryProtocol.new(transport)

  # Create a client to use the protocol encoder
  client = MovingAverageService::Client.new(protocol)

  current_time = (Time.now - start_time).round(5)
  puts "Creating a client:\t\t\t\t#{current_time}s"

  # Connect!
  start_time = Time.now
  transport.open
  current_time = (Time.now - start_time).round(5)
  puts "Opening connection:\t\t\t\t#{current_time}s"

  start_time = Time.now
  size = 384

  # Generate input
  data_in = Array.new(size)
  (0..(size - 1)).each do |i|
    data_in[i] = Float(rand(0..99))
  end
  current_time = (Time.now - start_time).round(5)
  puts "Generating input data:\t\t\t\t#{current_time}s"

  # Initialize maxfile
  start_time = Time.now
  max_file = client.MovingAverage_init()
  current_time = (Time.now - start_time).round(5)
  puts "Initializing maxfile:\t\t\t\t#{current_time}s"

  # Load DFE
  start_time = Time.now
  max_engine = client.max_load(max_file, '*')
  current_time = (Time.now - start_time).round(5)
  puts "Loading DFE:\t\t\t\t\t#{current_time}s"

  # Allocate and send input streams to server
  start_time = Time.now
  address_data_in = client.malloc_float(size)
  client.send_data_float(address_data_in, data_in)
  current_time = (Time.now - start_time).round(5)
  puts "Sending input data:\t\t\t\t#{current_time}s"

  # Allocate memory for output stream on server
  start_time = Time.now
  address_data_out = client.malloc_float(size)
  current_time = (Time.now - start_time).round(5)
  puts "Allocating memory for output stream on server:\t#{current_time}s"

  # Action default
  start_time = Time.now
  actions = MovingAverage_actions_t_struct.new
  actions.param_N = size
  actions.instream_x = address_data_in
  actions.outstream_y = address_data_out
  address_actions = client.send_MovingAverage_actions_t(actions)
  client.MovingAverage_run(max_engine, address_actions)
  current_time = (Time.now - start_time).round(5)
  puts "Moving average time:\t\t\t\t#{current_time}s"

  # Unload DFE
  start_time = Time.now
  client.max_unload(max_engine)
  current_time = (Time.now - start_time).round(5)
  puts "Unloading DFE:\t\t\t\t\t#{current_time}s"

  # Get output stream from server
  start_time = Time.now
  data_out = client.receive_data_float(address_data_out, size)
  current_time = (Time.now - start_time).round(5)
  puts "Getting output stream:\t(size = #{size * 32} bit)\t#{current_time}s"

  # Free allocated memory for streams on server
  start_time = Time.now
  client.free(address_data_in)
  client.free(address_data_out)
  client.free(address_actions)
  current_time = (Time.now - start_time).round(5)
  puts "Freeing allocated memory for streams on server:\t#{current_time}s"

  # Free allocated maxfile data
  start_time = Time.now
  client.MovingAverage_free()
  current_time = (Time.now - start_time).round(5)
  puts "Freeing allocated maxfile data:\t\t\t#{current_time}s"

  # Checking results
  start_time = Time.now
  status = check(data_in, data_out, size)
  current_time = (Time.now - start_time).round(5)
  puts "Checking results:\t\t\t\t#{current_time}s"

  # Close!
  start_time = Time.now
  transport.close
  current_time = (Time.now - start_time).round(5)
  puts "Closing connection:\t\t\t\t#{current_time}s"

  current_time = (Time.now - start_dfe_time).round(5)
  puts "DFE moving average total time:\t\t\t#{current_time}s"

  if status == 0
    puts 'Test successful!'
  else
    puts "Test failed #{status} times!"
    Kernel.exit(-1)
  end

rescue Thrift::Exception => tx
  print 'Thrift::Exception: ', tx.message, "\n"
  Kernel.exit(-1)
end

File: examples/MovingAverage/client/perl/AdvancedStatic/MovingAverageClient.pl
#!/usr/bin/perl -w

use strict;
use warnings;

use lib '../gen-perl';

use Thrift;
use Thrift::BinaryProtocol;
use Thrift::Socket;
use Thrift::BufferedTransport;

use com::maxeler::MovingAverage::MovingAverageService;
use com::maxeler::MovingAverage::Types;

use Data::Dumper;

use Time::HiRes qw/ time /;

sub check {
    my ($dataIn_ref, $dataOut_ref, $size) = @_;
    my @dataIn = @$dataIn_ref;
    my @dataOut = @$dataOut_ref;

    my $status = 0;

    for (my $i = 1; $i < $size - 1; $i++) {
        my $a = unpack('f', pack('f', $dataIn[$i - 1]));
        my $b = unpack('f', pack('f', $dataIn[$i]));
        my $c = unpack('f', pack('f', $dataIn[$i + 1]));
        my $res = unpack('f', pack('f', ($a + $b + $c) / 3));
        if ($dataOut[$i] != $res) {
            print "Output data @ $i = $dataOut[$i] (expected $res)\n";
            $status++;
        }
    }

    return $status;
}

my $startTime = time;
my $startDFETime = $startTime;

# Make socket
my $socket    = new Thrift::Socket('localhost', 9090);

# Buffering is critical. Raw sockets are very slow
my $transport = new Thrift::BufferedTransport($socket);

# Wrap in a protocol
my $protocol  = new Thrift::BinaryProtocol($transport);

# Create a client to use the protocol encoder
my $client    = new com::maxeler::MovingAverage::MovingAverageServiceClient($protocol);

print "Creating a client:\t\t\t\t", (time - $startTime), "s\n";

eval{
    # Connect!
    $startTime = time;
    $transport->open();
    print "Opening connection:\t\t\t\t", (time - $startTime), "s\n";

    $startTime = time;
    my $size = 384;
    # Generate two random vectors
    my @dataIn = ();
    for (my $i = 0; $i < $size; $i++) {
        $dataIn[$i] = int(rand(100));
    }
    print "Generating input data:\t\t\t\t", (time - $startTime), "s\n";

    # Initialize maxfile
    $startTime = time;
    my $max_file = $client->MovingAverage_init();
    print "Initializing maxfile:\t\t\t\t", (time - $startTime), "s\n";

    # Load DFE
    $startTime = time;
    my $max_engine = $client->max_load($max_file, '*');
    print "Loading DFE:\t\t\t\t\t", (time - $startTime), "s\n";

    # Allocate and send input streams to server
    $startTime = time;
    my $address_dataIn = $client->malloc_float($size);
    $client->send_data_float($address_dataIn, \@dataIn);
    print "Sending input data:\t\t\t\t", (time - $startTime), "s\n";

    # Allocate memory for output stream on server
    $startTime = time;
    my $address_dataOut = $client->malloc_float($size);
    print "Allocating memory for output stream on server:\t", (time - $startTime), "s\n";

    # Action default
    $startTime = time;
    my $actions = new com::maxeler::MovingAverage::MovingAverage_actions_t_struct();
    $actions->{param_N} = $size;
    $actions->{instream_x} = $address_dataIn;
    $actions->{outstream_y} = $address_dataOut;
    my $address_actions = $client->send_MovingAverage_actions_t($actions);
    $client->MovingAverage_run($max_engine, $address_actions);
    print "Moving average time:\t\t\t\t", (time - $startTime), "s\n";

    # Unload DFE
    $startTime = time;
    $client->max_unload($max_engine);
    print "Unloading DFE:\t\t\t\t\t", (time - $startTime), "s\n";

    # Get output stream from server
    $startTime = time;
    my $dataOut_ref = $client->receive_data_float($address_dataOut, $size);
    my @dataOut = @$dataOut_ref;
    print "Getting output stream:\t(size = ", ($size * 32), " bit)\t", (time - $startTime), "s\n";

    # Free allocated memory for streams on server
    $startTime = time;
    $client->free($address_dataIn);
    $client->free($address_dataOut);
    print "Freeing allocated memory for streams on server:\t", (time - $startTime), "s\n";

    # Free allocated maxfile data
    $startTime = time;
    $client->MovingAverage_free();
    print "Freeing allocated maxfile data:\t\t\t", (time - $startTime), "s\n";

    $startTime = time;
    my $status = check(\@dataIn, \@dataOut, $size);
    print "Checking results:\t\t\t\t", (time - $startTime), "s\n";
    
    # Close!
    $startTime = time;
    $transport->close();
    print "Closing connection:\t\t\t\t", (time - $startTime), "s\n";

    print "DFE moving average total time:\t\t\t", (time - $startDFETime), "s\n";

    if ($status == 0) { 
        print "Test successful!\n";
    } else {
        print "Test failed ", $status, " times!\n";
        exit -1;
    }

}; if($@){
   warn(Dumper($@));
}
